### PR TITLE
Add LLVM `minnum`/`maxnum` intrinsics

### LINKED ISSUE #18384: Use minnum/maxnum LLVM instrinsics for fmin/fmax
http://reviews.llvm.org/rL220341

Requires an LLVM upgrade, but not worth doing one just for this purpose.

### PR DESCRIPTION
**Pull request by varkor**
LLVM’s intrinsics `minnum` and `maxnum` are now used for `min` and
`max` on `f32` and `f64`. This resolves #18384.

### CODE REVIEW

**Comment by joshtriplett** [comment]
LGTM, as soon as CI passes.

File: src/libcore/intrinsics.rs
Comment (estebank) [inline src/libcore/intrinsics.rs]: Shouldn't you copy the deleted comments from `libcore` at least once here?

Comment (joshtriplett) [reply src/libcore/intrinsics.rs]: Good call. Even though these are only used for bootstrapping, they still ought to have that correctness warning.

**Review by joshtriplett — CHANGES_REQUESTED** [review]

File: src/libcore/intrinsics.rs
Comment (varkor) [reply src/libcore/intrinsics.rs]: Okay, I was wondering whether they were necessary if just used for bootstrapping. I'll add them back!

**Comment by kennytm** [comment]
@bors rollup- r-

These intrinsics (`llvm_minnum_f64`) are undefined on asm.js

**Comment by varkor** [comment]
It looks like in order to get these working with asm.js, I'll probably need to create a PR on the emscripten repository to implement them (I'm not sure how quick the turnaround is after a PR is merged on the repository — any idea what the process is after that @kennytm?). I'll try to do this soon.

**Comment by scottmcm** [comment]
From https://github.com/rust-lang/rust/issues/42423#issuecomment-306053932
> There is an LLVM intrinsic llvm.minnum.fNN. When compared with the pure Rust implementation, the generated assembly is generally worse on x86_64 and i686, and better on ARM and AArch64 (Both won't call fmaxf/fminf unless there is no hard-float support). 

Is the x64 assembly better now?

**Comment by rkruppe** [comment]
Running the test case from the linked comment through godbolt reveals that the x86 situation is still similar: identical asm for general min/max, but the intrinsic optimizes less well when used for clamping.

**Comment by varkor** [comment]
This is quite interesting — before submitting the PR, I checked that the intrinsics generated better instructions in isolation than the existing Rust implementation, but in @rkruppe's example, they're identical. [Here's a very small change to the example](https://godbolt.org/g/GHZB8x) — the only thing that's changed is the order of the conditions, but the generated asm is significantly worse.

```rust
if other > self || other.is_nan() { self } else { other } // 16 instructions (min)
if other.is_nan() || other > self { self } else { other } // 11 instructions (min)
```
`max` is similar. Notice that it's the former, less efficient, version which is currently used in the standard library.

I hadn't spotted the issue shown by clamping, but given that the advantage the intrinsics demonstrate seems to be achievable by the Rust code, I imagine it'd be better to hold off from using the intrinsics for `min` and `max` for now.

What's the best way to proceed? Use Rust for `min` and `max`, but make the condition optimisation here? Is it worth having the intrinsics if they're not directly used in the standard library yet?

**Comment by rkruppe** [comment]
Optimizing the Rust implementation would be great. I don't think there's any point in having the intrinsics if we're not going to use them. Very unfortunate that you've already spent the time on them, but at least another optimization came out of it :)

**Comment by varkor** [comment]
Let's close this PR for the time being, then. When LLVM gets a little bit better at optimising `minnum`/`maxnum` intrinsics we can look into this again.

**Pull request closed without merging**
